Add docker-compose file to work with ES

sergeyklay opened the pull request:
## Which problem is this PR solving?
Resolves https://github.com/jaegertracing/jaeger/issues/931

## Short description of the changes
This PR aims to reduce misunderstanding on running Jaeger with ES using docker-compose.

pavolloffay commented:
@seregayoga thanks for the PR. 

Some time ago I had this idea https://github.com/jaegertracing/jaeger/issues/502. Basically we could have 3 compose files, one for ES, one for cassanra and one generic with only jaeger services. Do you think you can make it work?

Other thing, there should be probably a test to verify that compose works. Something like https://github.com/jaegertracing/jaeger/issues/405 should help.

jpkrohling commented on docker-compose/jaeger-docker-compose-es.yml:
```diff
@@ -0,0 +1,99 @@
+      - jaeger-collector
+
+  jaeger-query:
+    image: jaegertracing/jaeger-query
```
Wouldn't it be better to pin to a specific version?

jpkrohling commented on docker-compose/jaeger-docker-compose-es.yml:
```diff
@@ -0,0 +1,99 @@
+    command: [
+      "--es.server-urls=http://elasticsearch:9200",
+      "--span-storage.type=elasticsearch",
+      "--log-level=error"
```
I don't think it should be set, unless there's a good reason to default to `error` for the `docker-compose` but not in all other deployment options (like, [Kubernetes](https://github.com/jaegertracing/jaeger-kubernetes/blob/master/production-elasticsearch/configmap.yml#L32) or [OpenShift](https://github.com/jaegertracing/jaeger-openshift/blob/master/production/configmap-elasticsearch.yml#L30))

sergeyklay commented:
@pavolloffay Sounds good to me. Let's just use step-by-step changes (feature per PR).

yurishkuro commented:
> but I'm not sure this one here from this PR is being tested at all.

that is my main concern as well - this is a file that's not being tested anywhere.

sergeyklay commented:
What do you think about [RSpec](http://rspec.info) and [ServerSpec](https://serverspec.org)? They use Ruby but the syntax is extremely simple. For example this tests that we have Ubuntu 14 installed:

```ruby
# spec/Dockerfile_spec.rb

require "serverspec"
require "docker"

describe "Dockerfile" do
  before(:all) do
    image = Docker::Image.build_from_dir('.')

    set :os, family: :debian
    set :backend, :docker
    set :docker_image, image.id
  end

  it "installs the right version of Ubuntu" do
    expect(os_version).to include("Ubuntu 14")
  end

  def os_version
    command("lsb_release -a").stdout
  end
end
```

jpkrohling commented:
> What do you think about RSpec and ServerSpec?

We have tests already using another [`docker-compose`](https://github.com/jaegertracing/jaeger/blob/master/docker-compose/jaeger-docker-compose.yml) file: 

https://github.com/jaegertracing/jaeger/blob/master/crossdock/rules.mk#L2

sergeyklay commented:
@jpkrohling Should I make `JAEGER_COMPOSE_YAML` configurable at testing stage?

jpkrohling commented:
> Should I make JAEGER_COMPOSE_YAML configurable at testing stage?

That would be wonderful, yes! Take a look also at the Travis configuration, it would probably make sense to add a new matrix element, with `DOCKER_COMPOSE=true` and the new env var to control `JAEGER_COMPOSE_YAML`.

sergeyklay commented:
@yurishkuro @jpkrohling @pavolloffay I've modified current crossdock tests so that there is an ability to run tests as follows:

#### Default behavior
```shell
# docker-compose/jaeger-docker-compose-cassandra.yml
make crossdock
```

#### Using Cassandra as Storage Backend
```shell
# docker-compose/jaeger-docker-compose-cassandra.yml
export JAEGER_STORAGE_BACKEND=cassandra
make crossdock
```

#### Using Elasticsearch as Storage Backend
```shell
# docker-compose/jaeger-docker-compose-elasticsearch.yml
export JAEGER_STORAGE_BACKEND=elasticsearch
make crossdock
```

However, unfortunately the _latest several_ tests for Elasticsearch failed:
```
2018/07/16 21:05:55 HTTP: HEAD http://test_driver:8080/
2018/07/16 21:05:55 HTTP: Service test_driver:8080 timed out. Last error: Head http://test_driver:8080/: dial tcp: lookup test_driver on 127.0.0.11:53: no such host
2018/07/16 21:05:55 HTTP: HEAD http://java:8080/
2018/07/16 21:05:55 HTTP: Service java:8080 timed out. Last error: Head http://java:8080/: dial tcp: lookup java on 127.0.0.11:53: no such host
2018/07/16 21:05:55 Error: One or more services timed out after 4m0s
```

There is also a lot of docker logs like:
```
/home/travis/gopath/src/github.com/jaegertracing/jaeger/crossdock/main.go:66
WARN crossdock/main.go:96 Health check failed
{
    "service": "Query",
    "error": "Get http://jaeger-query:16687: dial tcp: lookup jaeger-query on 127.0.0.11:53: no such host"
}
```

I ve increased minimum and maximum heap size to 1 GB by passing environment variable `ES_JAVA_OPTS=-Xms1g -Xmx1g` to the elasticsearch container as described in the https://www.elastic.co/guide/en/elasticsearch/reference/6.3/heap-size.html but still no luck.

So my questions is:
1. Is this related to the Elasticsearch either my PR?
2. Should we use `WARN` log level for errors like "no such host"
3. I noticed that the Elasticsearch tests do not finish so quickly as the same using Cassandra. Is this related to the Elasticsearch and Cassandra faster by 30% or it is Jaeger implementation?

<img width="1160" alt="screen shot 2018-07-17 at 12 29 42 am" src="https://user-images.githubusercontent.com/1256298/42784770-e0ca3f9e-8958-11e8-9adb-5fcd3a534247.png">

Also I noticed that we collect logs:

```shell
docker-compose -f $(JAEGER_COMPOSE_YAML) -f $(XDOCK_YAML) run crossdock 2>&1 | tee run-crossdock.log
grep 'Tests passed!' run-crossdock.log
```

But never read this file if build fails. Do we really need this log file? For example we can just return -1 instead of grep-ing `Tests passed`

Finally, any idea how to fix tests?

----

Note: I've renamed current docker-compose file:
```diff
- docker-compose/jaeger-docker-compose.yml
+ docker-compose/jaeger-docker-compose-cassandra.yml
```

black-adder commented:
Has it been consistently failing for:
```
✗ [endtoend] test_driver→ (services=zipkin-brave-json) ⇒ Fail: could not retrieve traces from query service
✗ [endtoend] test_driver→ (services=zipkin-brave-json-v2) ⇒ Fail: could not retrieve traces from query service
✗ [endtoend] test_driver→ (services=zipkin-brave-thrift) ⇒ Fail: could not retrieve traces from query service
```  like in the most recent run or have there been multiple issues?

sergeyklay commented:
@black-adder Yes, just saw it. These are the new errors (after ead2abf) that we saw because the previous tests have passed.

sergeyklay commented:
Well, now logs says:
```
test_driver_1 | 2018-07-16T22:34:34.692Z INFO services/tracehandler.go:95
	Starting EndToEnd test	{"service": "zipkin-brave-json-v2"}

test_driver_1 | 2018-07-16T22:34:34.975Z INFO services/tracehandler.go:224
	Waiting for traces, iteration 1 out of 10

test_driver_1 | 2018-07-16T22:34:34.981Z INFO services/query.go:74
	Retrieved trace from query	{"body": "{\"data\":[],\"total\":0,\"limit\":0,\"offset\":0,\"errors\":null}", "url": "http://jaeger-query:16686/api/traces?end=1531780474000000&operation=4dba7b0f9da1d7eb&service=crossdock-zipkin-brave-json-v2&tag=7cd4b7a55a25e0cb%3Aa2b894cf840ec4b"}

test_driver_1 | 2018-07-16T22:34:34.981Z INFO services/tracehandler.go:229
	Could not retrieve trace from query service
```

I still have no idea why "Could not retrieve trace from query service" treats as INFO. But anyway tests failed:
```
✓ [endtoend] test_driver→ (services=go) ⇒ Pass
✓ [endtoend] test_driver→ (services=java) ⇒ Pass
✓ [endtoend] test_driver→ (services=node) ⇒ Pass
✓ [endtoend] test_driver→ (services=python) ⇒ Pass
✗ [endtoend] test_driver→ (services=zipkin-brave-json) ⇒ Fail: could not retrieve traces from query service
✗ [endtoend] test_driver→ (services=zipkin-brave-json-v2) ⇒ Fail: could not retrieve traces from query service
✗ [endtoend] test_driver→ (services=zipkin-brave-thrift) ⇒ Fail: could not retrieve traces from query service
```

Does any one tested this with elasticsearch docker-compose at least on localhost? :)

black-adder commented:
nvm, just checked build history, looks like the tests never even started

sergeyklay commented:
@black-adder Any news? :)

sergeyklay commented:
Any news?

sergeyklay commented:
Any chance to get PR's priority?
